bindings/rust/src/lib.rs

```diff
@@ -3,16 +3,233 @@
+    pub fn value(&self) -> Option<Value> {
+        if !self.0.trapped && self.0.has_value {
+            Some(self.0.value)
+        } else {
+            None
+        }
+    }
+}
+
+impl From<ExecutionResult> for sys::FizzyExecutionResult {
+    fn from(v: ExecutionResult) -> Self {
+        v.0
+    }
+}
+
+impl Instance {
+    /// Get a read-only pointer to the module.
+    unsafe fn get_instance_module(&self) -> *const sys::FizzyModule {
+        sys::fizzy_get_instance_module(self.0.as_ptr())
+    }
+
+    /// Find index of exported function by name.
+    pub fn find_exported_function(&self, name: &str) -> Option<u32> {
+        let module = unsafe { sys::fizzy_get_instance_module(self.0.as_ptr()) };
+        let name = CString::new(name).expect("CString::new failed");
+        let mut func_idx: u32 = 0;
+        let found =
+            unsafe { sys::fizzy_find_exported_function(module, name.as_ptr(), &mut func_idx) };
+        if found {
+            Some(func_idx)
+        } else {
+            None
+        }
+    }
+
+    /// Unsafe execution of a given function index `func_idx` with the given values `args`.
+    ///
+    /// An invalid index, invalid inputs, or invalid depth can cause undefined behaviour.
+    ///
+    /// The `depth` argument sets the initial call depth. Should be set to `0`.
+    ///
+    /// # Safety
+    /// This function expects a valid `func_idx`, appropriate number of `args`, and valid `depth`.
+    pub unsafe fn unsafe_execute(
+        &mut self,
+        func_idx: u32,
+        args: &[Value],
+        depth: i32,
+    ) -> ExecutionResult {
+        ExecutionResult {
+            0: sys::fizzy_execute(self.0.as_ptr(), func_idx, args.as_ptr(), depth),
+        }
+    }
+
+    /// Find function type for a given index. Must be a valid index otherwise behaviour is undefine.
+    unsafe fn get_function_type(&self, func_idx: u32) -> sys::FizzyFunctionType {
+        let module = self.get_instance_module();
+        sys::fizzy_get_function_type(module, func_idx)
+    }
+
+    /// Execute a given function of `name` with the given values `args`.
+    ///
+    /// An error is returned if the function can not be found, or inappropriate number of arguments are passed.
+    /// However the types of the arguments are not validated.
+    ///
+    /// The `depth` argument sets the initial call depth. Should be set to `0`.
+    pub fn execute(
+        &mut self,
+        name: &str,
+        args: &[Value],
+        depth: i32,
+    ) -> Result<ExecutionResult, ()> {
+        let func_idx = self.find_exported_function(&name);
+        if func_idx.is_none() {
+            return Err(());
+        }
+        let func_idx = func_idx.unwrap();
+
+        let func_type = unsafe { self.get_function_type(func_idx) };
+        if func_type.inputs_size != args.len() {
+            return Err(());
+        }
+        // TODO: validate input types too
+
+        Ok(unsafe { self.unsafe_execute(func_idx, &[], depth) })
```

Comment:
```suggestion
        Ok(unsafe { self.unsafe_execute(func_idx, args, depth) })
```

Reply:
Oh wow, need to improve the unit tests.

Reply:
This functionality is moved to the next PR.